### Goal:
### 1. Will a fellow get placed or not
### 2. How long will it take to get placed if placed

### Part 1: Fellow will get placed or not

In [2]:
import numpy as np
import pandas as pd
from keras.utils import to_categorical

Using TensorFlow backend.


In [3]:
data = pd.read_csv("data_Pathrise.csv")

In [4]:
data.head()

id pathrise_status primary_track cohort_tag  program_duration_days  placed  \
0   1          Active           SWE     OCT19A                    NaN       0   
1   2          Active           PSO     JAN20A                    NaN       0   
2   3     Closed Lost        Design     AUG19B                    0.0       0   
3   4     Closed Lost           PSO     AUG19B                    0.0       0   
4   5          Placed           SWE     AUG19A                   89.0       1   

   employment_status  highest_level_of_education length_of_job_search  \
0          Unemployed          Bachelor's Degree           3-5 months   
1          Unemployed    Some College, No Degree           3-5 months   
2  Employed Part-Time            Master's Degree  Less than one month   
3          Contractor          Bachelor's Degree  Less than one month   
4          Unemployed          Bachelor's Degree           1-2 months   

            biggest_challenge_in_search professional_experience  \
0       Hearing back on my applications               3-4 years   
1   Getting past final round interviews               1-2 years   
2  Figuring out which jobs to apply for      Less than one year   
3   Getting past final round interviews      Less than one year   
4       Hearing back on my applications               1-2 years   

  work_authorization_status  number_of_interviews  number_of_applications  \
0            Canada Citizen                   2.0                     900   
1                   Citizen                   6.0                       0   
2                   Citizen                   0.0                       0   
3                   Citizen                   5.0                      25   
4               F1 Visa/OPT                  10.0                     100   

  gender                                 race  
0   Male  Non-Hispanic White or Euro-American  
1   Male  Non-Hispanic White or Euro-American  
2   Male         East Asian or Asian American  
3   Male             Decline to Self Identify  
4   Male         East Asian or Asian American

In [5]:
headings_list = list(data.columns)
headings_list

['id',
 'pathrise_status',
 'primary_track',
 'cohort_tag',
 'program_duration_days',
 'placed',
 'employment_status ',
 'highest_level_of_education',
 'length_of_job_search',
 'biggest_challenge_in_search',
 'professional_experience',
 'work_authorization_status',
 'number_of_interviews',
 'number_of_applications',
 'gender',
 'race']

In [6]:
#Checking whether all ids are unique
data[data['id'].duplicated() == True]

Empty DataFrame
Columns: [id, pathrise_status, primary_track, cohort_tag, program_duration_days, placed, employment_status , highest_level_of_education, length_of_job_search, biggest_challenge_in_search, professional_experience, work_authorization_status, number_of_interviews, number_of_applications, gender, race]
Index: []

#### Converting features with text to numbers so that it can be fed to an ML algorithm

In [7]:
headings_list_for_num_process = headings_list[:]
headings_list_for_num_process.remove('id')
headings_list_for_num_process.remove('placed')
headings_list_for_num_process.remove('program_duration_days')
headings_list_for_num_process.remove('number_of_interviews')
headings_list_for_num_process.remove('number_of_applications')

In [8]:
#Below are the features that are going to be converted to multiple columns 
# with numeric values based on the text information provided.Although cohort_tag 
# might not be an important feature, I'm including it now.
headings_list_for_num_process

['pathrise_status',
 'primary_track',
 'cohort_tag',
 'employment_status ',
 'highest_level_of_education',
 'length_of_job_search',
 'biggest_challenge_in_search',
 'professional_experience',
 'work_authorization_status',
 'gender',
 'race']

In [9]:
#creating unique columns dictionary
new_dataframe_dict = {}
all_uniques_encoded = {}
for i in headings_list_for_num_process:
    codes, uniques = pd.factorize(data[i])
    new_dataframe_dict[i] = codes
    all_uniques_encoded[i] = uniques

In [10]:
all_uniques_encoded

{'pathrise_status': Index(['Active', 'Closed Lost', 'Placed', 'Withdrawn (Failed)',
        'Withdrawn (Trial)', 'Withdrawn', 'Deferred', 'Break', 'MIA'],
       dtype='object'),
 'primary_track': Index(['SWE', 'PSO', 'Design', 'Data', 'Web', 'Marketing'], dtype='object'),
 'cohort_tag': Index(['OCT19A', 'JAN20A', 'AUG19B', 'AUG19A', 'SEP19A', 'AUG19C', 'DEC19A',
        'FEB20A', 'NOV19A', 'NOV19B', 'JAN20B', 'FEB20B', 'JUL19B', 'APR20A',
        'OCT18A', 'SEP18C', 'OCT18B', 'JAN19A', 'MAR19A', 'FEB19A', 'FEB19B',
        'APR19A', 'AUG18A', 'DEC18A', 'MAR19B', 'JAN19B', 'JUN19B', 'SEP18B',
        'NOV18A', 'SEP18A', 'APR19B', 'MAY19A', 'JUL19A', 'JUN19A', 'FEB18A',
        'JUN18A', 'MAR18A', 'APR18A', 'APR18B', 'MAY18A', 'OCT21A', 'SEP19B',
        'MAR20A', 'MAY19B', 'FEB20a', 'OCT19B', 'JAN18A'],
       dtype='object'),
 'employment_status ': Index(['Unemployed', 'Employed Part-Time', 'Contractor', 'Employed Full-Time',
        'Student'],
       dtype='object'),
 'highest_level

In [11]:
#Creating new columns based on the text information
df_X = pd.DataFrame()
for i in headings_list_for_num_process:
    for j in all_uniques_encoded[i]:
        df_X[i+'_'+j] = [1 if x == j else 0 for x in data[i]]     

In [12]:
df_X.shape

(2544, 115)

In [14]:
df_X.head()

pathrise_status_Active  pathrise_status_Closed Lost  \
0                       1                            0   
1                       1                            0   
2                       0                            1   
3                       0                            1   
4                       0                            0   

   pathrise_status_Placed  pathrise_status_Withdrawn (Failed)  \
0                       0                                   0   
1                       0                                   0   
2                       0                                   0   
3                       0                                   0   
4                       1                                   0   

   pathrise_status_Withdrawn (Trial)  pathrise_status_Withdrawn  \
0                                  0                          0   
1                                  0                          0   
2                                  0                          0   
3                                  0                          0   
4                                  0                          0   

   pathrise_status_Deferred  pathrise_status_Break  pathrise_status_MIA  \
0                         0                      0                    0   
1                         0                      0                    0   
2                         0                      0                    0   
3                         0                      0                    0   
4                         0                      0                    0   

   primary_track_SWE  ...  gender_Non-Binary  \
0                  1  ...                  0   
1                  0  ...                  0   
2                  0  ...                  0   
3                  0  ...                  0   
4                  1  ...                  0   

   race_Non-Hispanic White or Euro-American  \
0                                         1   
1                                         1   
2                                         0   
3                                         0   
4                                         0   

   race_East Asian or Asian American  race_Decline to Self Identify  \
0                                  0                              0   
1                                  0                              0   
2                                  1                              0   
3                                  0                              1   
4                                  1                              0   

   race_Black, Afro-Caribbean, or African American  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   

   race_Latino or Hispanic American  race_Middle Eastern or Arab American  \
0                                 0                                     0   
1                                 0                                     0   
2                                 0                                     0   
3                                 0                                     0   
4                                 0                                     0   

   race_South Asian or Indian American  race_Two or More Races  \
0                                    0                       0   
1                                    0                       0   
2                                    0                       0   
3                                    0                       0   
4                                    0                       0   

   race_Native American or Alaskan Native  
0                                       0  
1                                       0  
2                                       0  
3                                       0  
4    

In [15]:
#adding the features that need not be modified for analysis
df_X['program_duration_days'] = data['program_duration_days']
df_X['number_of_interviews'] = data['number_of_interviews']
df_X['number_of_applications'] = data['number_of_applications']

In [16]:
df_X = df_X.fillna(0)

In [17]:
df_X.head()

pathrise_status_Active  pathrise_status_Closed Lost  \
0                       1                            0   
1                       1                            0   
2                       0                            1   
3                       0                            1   
4                       0                            0   

   pathrise_status_Placed  pathrise_status_Withdrawn (Failed)  \
0                       0                                   0   
1                       0                                   0   
2                       0                                   0   
3                       0                                   0   
4                       1                                   0   

   pathrise_status_Withdrawn (Trial)  pathrise_status_Withdrawn  \
0                                  0                          0   
1                                  0                          0   
2                                  0                          0   
3                                  0                          0   
4                                  0                          0   

   pathrise_status_Deferred  pathrise_status_Break  pathrise_status_MIA  \
0                         0                      0                    0   
1                         0                      0                    0   
2                         0                      0                    0   
3                         0                      0                    0   
4                         0                      0                    0   

   primary_track_SWE  ...  race_Decline to Self Identify  \
0                  1  ...                              0   
1                  0  ...                              0   
2                  0  ...                              0   
3                  0  ...                              1   
4                  1  ...                              0   

   race_Black, Afro-Caribbean, or African American  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   

   race_Latino or Hispanic American  race_Middle Eastern or Arab American  \
0                                 0                                     0   
1                                 0                                     0   
2                                 0                                     0   
3                                 0                                     0   
4                                 0                                     0   

   race_South Asian or Indian American  race_Two or More Races  \
0                                    0                       0   
1                                    0                       0   
2                                    0                       0   
3                                    0                       0   
4                                    0                       0   

   race_Native American or Alaskan Native  program_duration_days  \
0                                       0                    0.0   
1                                       0                    0.0   
2                                       0                    0.0   
3                                       0                    0.0   
4                                       0                   89.0   

   number_of_interviews  number_of_applications  
0                   2.0                     900  
1                   6.0                       0  
2                   0.0                       0  
3                   5.0                      25  
4                  10.0                     100  

[5 rows x 118 columns]

In [18]:
df_Y = data['placed']

In [20]:
#Converting pandas dataframe to numpy for sklearn and easier manipulation purposes
X = df_X.to_numpy()
Y = df_Y.to_numpy()

In [22]:
X.shape

(2544, 118)

In [23]:
Y.shape

(2544,)

In [24]:
#Splitting data to test and train sets
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=3)
print('X_train_shape = ', X_train.shape)
print('X_test_shape = ', X_test.shape)

X_train_shape =  (2289, 118)
X_test_shape =  (255, 118)


In [25]:
#Convert to categorical
y_train_cat = to_categorical(Y_train)
y_test_cat = to_categorical(Y_test)

print('y_train_cat shape = ', y_train_cat.shape)
print('y_test_cat shape = ', y_test_cat.shape)

y_train_cat shape =  (2289, 2)
y_test_cat shape =  (255, 2)


In [197]:
import warnings
warnings.filterwarnings("ignore")  # Ignore some warning logs


from keras.models import Sequential
from keras.layers import Dense, Dropout


#  Define a Feed-Forward Model with 2 hidden layers with dimensions 392 and 196 Neurons
model_1 = Sequential([
  Dense(100, activation='relu', input_shape=(118,), name="first_hidden_layer"),
  Dense(100, activation='relu', name="second_hidden_layer"), 
  Dense(100,activation= 'relu',name = "third_hidden_layer"),
  Dense(100,activation= 'relu',name = "forth_hidden_layer"),
  Dense(100,activation= 'relu',name = "fifth_hidden_layer"),
  Dense(2, activation='softmax'),
])

#  Validate your Model Architecture
print(model_1.summary())

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
first_hidden_layer (Dense)   (None, 100)               11900     
_________________________________________________________________
second_hidden_layer (Dense)  (None, 100)               10100     
_________________________________________________________________
third_hidden_layer (Dense)   (None, 100)               10100     
_________________________________________________________________
forth_hidden_layer (Dense)   (None, 100)               10100     
_________________________________________________________________
fifth_hidden_layer (Dense)   (None, 100)               10100     
_________________________________________________________________
dense_5 (Dense)              (None, 2)                 202       
Total params: 52,502
Trainable params: 52,502
Non-trainable params: 0
__________________________________________________

In [200]:
import keras_metrics as km
# Compile model
model_1.compile(optimizer='sgd', loss='categorical_crossentropy',metrics=['accuracy',km.binary_precision(), km.binary_recall()])

# Train model
model_1.fit(X_train, y_train_cat, epochs=250, batch_size=10,)

tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> tp
tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> fp
tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> tp
tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> fn
Epoch 1/250
2289/2289 [==============================] - 2s 692us/step - loss: 0.2270 - accuracy: 0.8995 - precision: 0.9538 - recall: 0.8998
Epoch 2/250
2289/2289 [==============================] - 1s 341us/step - loss: 0.2513 - accuracy: 0.8895 - precision: 0.8956 - recall: 0.9331
Epoch 3/250
2289/2289 [==============================] - 1s 338us/step - loss: 0.2263 - accuracy: 0.8995 - precision: 0.9446 - recall: 0.9219
Epoch 4/250
2289/2289 [==============================] - 1s 345us/step - loss: 0.2656 - accuracy: 0.8917 - precision: 0.9419 - recall: 0.8992
Epoch 5/250
2289/2289 [==============================] - 1s 364us/step - loss: 0.2547 - accuracy: 0.8807 - precision: 0.9306 - recall: 0.8691
Epoch 6/250
22

2289/2289 [==============================] - 1s 348us/step - loss: 0.2452 - accuracy: 0.8764 - precision: 0.9397 - recall: 0.8614
Epoch 57/250
2289/2289 [==============================] - 1s 482us/step - loss: 0.2513 - accuracy: 0.8860 - precision: 0.9490 - recall: 0.8704
Epoch 58/250
2289/2289 [==============================] - 1s 543us/step - loss: 0.2747 - accuracy: 0.8606 - precision: 0.9483 - recall: 0.7706
Epoch 59/250
2289/2289 [==============================] - 1s 379us/step - loss: 0.2226 - accuracy: 0.8938 - precision: 0.9236 - recall: 0.9218
Epoch 60/250
2289/2289 [==============================] - 1s 350us/step - loss: 0.2635 - accuracy: 0.8790 - precision: 0.9454 - recall: 0.8896
Epoch 61/250
2289/2289 [==============================] - 1s 345us/step - loss: 0.3068 - accuracy: 0.8419 - precision: 0.9017 - recall: 0.7983
Epoch 62/250
2289/2289 [==============================] - 1s 346us/step - loss: 0.2720 - accuracy: 0.8576 - precision: 0.8997 - recall: 0.8837
Epoch 63/250

2289/2289 [==============================] - 1s 347us/step - loss: 0.2717 - accuracy: 0.8807 - precision: 0.9263 - recall: 0.8582
Epoch 113/250
2289/2289 [==============================] - 1s 352us/step - loss: 0.2327 - accuracy: 0.8991 - precision: 0.9606 - recall: 0.8689
Epoch 114/250
2289/2289 [==============================] - 1s 359us/step - loss: 0.2066 - accuracy: 0.9131 - precision: 0.9527 - recall: 0.9229
Epoch 115/250
2289/2289 [==============================] - 1s 364us/step - loss: 0.2272 - accuracy: 0.9013 - precision: 0.9573 - recall: 0.9179
Epoch 116/250
2289/2289 [==============================] - 1s 366us/step - loss: 0.2618 - accuracy: 0.8724 - precision: 0.9529 - recall: 0.8515
Epoch 117/250
2289/2289 [==============================] - 1s 342us/step - loss: 0.2351 - accuracy: 0.8943 - precision: 0.9698 - recall: 0.8676
Epoch 118/250
2289/2289 [==============================] - 1s 357us/step - loss: 0.2211 - accuracy: 0.8917 - precision: 0.9392 - recall: 0.8490
Epoch 

2289/2289 [==============================] - 1s 352us/step - loss: 0.2703 - accuracy: 0.8685 - precision: 0.9116 - recall: 0.8926
Epoch 168/250
2289/2289 [==============================] - 1s 360us/step - loss: 0.2611 - accuracy: 0.8742 - precision: 0.9548 - recall: 0.8694
Epoch 169/250
2289/2289 [==============================] - 1s 347us/step - loss: 0.2604 - accuracy: 0.8724 - precision: 0.9334 - recall: 0.8667
Epoch 170/250
2289/2289 [==============================] - 1s 367us/step - loss: 0.2381 - accuracy: 0.9017 - precision: 0.9648 - recall: 0.8955
Epoch 171/250
2289/2289 [==============================] - 1s 361us/step - loss: 0.2726 - accuracy: 0.8777 - precision: 0.9290 - recall: 0.8661
Epoch 172/250
2289/2289 [==============================] - 1s 310us/step - loss: 0.2275 - accuracy: 0.8921 - precision: 0.9609 - recall: 0.8543
Epoch 173/250
2289/2289 [==============================] - 1s 345us/step - loss: 0.2364 - accuracy: 0.9000 - precision: 0.9324 - recall: 0.9340
Epoch 

2289/2289 [==============================] - 1s 325us/step - loss: 0.2090 - accuracy: 0.9065 - precision: 0.9430 - recall: 0.9122
Epoch 224/250
2289/2289 [==============================] - 1s 351us/step - loss: 0.2329 - accuracy: 0.8925 - precision: 0.9304 - recall: 0.9054
Epoch 225/250
2289/2289 [==============================] - 1s 340us/step - loss: 0.2300 - accuracy: 0.9104 - precision: 0.9311 - recall: 0.9211
Epoch 226/250
2289/2289 [==============================] - 1s 352us/step - loss: 0.2511 - accuracy: 0.8781 - precision: 0.8913 - recall: 0.8960
Epoch 227/250
2289/2289 [==============================] - 1s 357us/step - loss: 0.2428 - accuracy: 0.8834 - precision: 0.9253 - recall: 0.8738
Epoch 228/250
2289/2289 [==============================] - 1s 334us/step - loss: 0.2478 - accuracy: 0.8847 - precision: 0.9584 - recall: 0.8637
Epoch 229/250
2289/2289 [==============================] - 1s 337us/step - loss: 0.3446 - accuracy: 0.8292 - precision: 0.9442 - recall: 0.7587
Epoch 

### The above model is trained with training accuracy  89%.  

In [201]:
# Evaluate your model's performance on the test data
performance_test = model_1.evaluate(X_test, y_test_cat)
print("Accuracy on Test samples: {0}".format(performance_test[1]))

255/255 [==============================] - 0s 1ms/step
Accuracy on Test samples: 0.9764705896377563


### The model predicts whether a fellow will be placed oe not with 97% accuracy. Precision and recall are close to 1. This means that the model is relevant

In [202]:
#An accuracy calculating function

def accuracy(y,Y_pred):
  count = 0
  for i,val in enumerate(y):
    if (val == Y_pred[i]):
      count = count + 1
  accuracy = count/y.shape[0]
  return accuracy

### Using decision tree algorithm

In [212]:
from sklearn import tree
clf = tree.DecisionTreeClassifier(max_depth = 50)
clf = clf.fit(X_train, Y_train)

In [213]:
y_pred_test = clf.predict(X_test)
print("Accuracy with decision tree approach for test set = ",accuracy(Y_test,y_pred_test))

Accuracy with decision tree approach for test set =  1.0


### Part 2: If placed, how long will it take?

In [214]:
#Filtering to get data of those candidates placed
df_X_new = df_X[df_Y==1]

In [215]:
df_X_new.shape

(956, 118)

In [216]:
df_X_new.head()

pathrise_status_Active  pathrise_status_Closed Lost  \
4                        0                            0   
23                       0                            0   
30                       0                            0   
39                       0                            0   
54                       0                            0   

    pathrise_status_Placed  pathrise_status_Withdrawn (Failed)  \
4                        1                                   0   
23                       1                                   0   
30                       1                                   0   
39                       1                                   0   
54                       1                                   0   

    pathrise_status_Withdrawn (Trial)  pathrise_status_Withdrawn  \
4                                   0                          0   
23                                  0                          0   
30                                  0                          0   
39                                  0                          0   
54                                  0                          0   

    pathrise_status_Deferred  pathrise_status_Break  pathrise_status_MIA  \
4                          0                      0                    0   
23                         0                      0                    0   
30                         0                      0                    0   
39                         0                      0                    0   
54                         0                      0                    0   

    primary_track_SWE  ...  race_Decline to Self Identify  \
4                   1  ...                              0   
23                  0  ...                              0   
30                  1  ...                              0   
39                  1  ...                              0   
54                  1  ...                              0   

    race_Black, Afro-Caribbean, or African American  \
4                                                 0   
23                                                0   
30                                                0   
39                                                0   
54                                                0   

    race_Latino or Hispanic American  race_Middle Eastern or Arab American  \
4                                  0                                     0   
23                                 0                                     0   
30                                 0                                     0   
39                                 0                                     0   
54                                 0                                     0   

    race_South Asian or Indian American  race_Two or More Races  \
4                                     0                       0   
23                                    0                       0   
30                                    0                       0   
39                                    0                       0   
54                                    0                       0   

    race_Native American or Alaskan Native  program_duration_days  \
4                                        0                   89.0   
23                                       0                  193.0   
30                                       0                   73.0   
39                                       1                   83.0   
54                                       0                   76.0   

    number_of_interviews  number_of_applications  
4                   10.0                     100  
23                   5.0                       4  
30                   0.0                       1  
39                   0.0                      15  
54                   0.0                      30  

[5 rows x 118 columns]

In [217]:
#length of job search is the interested result. So, removing all the columns related to that and this
# is a classification problem with 5 classes with 3-5 months as class 0, less than one month as class 1 and so on.
for i in all_uniques_encoded['length_of_job_search']:
    df_X_new = df_X_new.drop('length_of_job_search_'+i, axis=1)
    print(i)

3-5 months
Less than one month
1-2 months
6 months to a year
Over a year


In [218]:
df_X_new.shape

(956, 113)

In [219]:
#Converting text to obtain class numbers such as 0,1,2,3 and 4
df_Y_new = data.groupby(['length_of_job_search']).ngroup()

In [220]:
len(df_Y_new)

2544

In [221]:
print(max(df_Y_new))

4


In [222]:
#Removing rows of those that were not placed
df_Y_new = df_Y_new[df_Y==1]

In [223]:
df_Y_new.shape

(956,)

In [224]:
X_placed = df_X_new.to_numpy()

In [225]:
Y_placed = df_Y_new.to_numpy()

In [226]:
X_train_placed, X_test_placed, Y_train_placed, Y_test_placed = train_test_split(X_placed, Y_placed, test_size=0.1, random_state=4)
print('X_train_placed_shape = ', X_train_placed.shape)
print('X_test_placed_shape = ', X_test_placed.shape)

X_train_placed_shape =  (860, 113)
X_test_placed_shape =  (96, 113)


In [228]:
y_train_placed_cat = to_categorical(Y_train_placed)
y_test_placed_cat = to_categorical(Y_test_placed)

print('y_train_placed_cat shape = ', y_train_placed_cat.shape)
print('y_test_placed_cat shape = ', y_test_placed_cat.shape)

y_train_placed_cat shape =  (860, 5)
y_test_placed_cat shape =  (96, 5)


In [232]:
import warnings
warnings.filterwarnings("ignore")  # Ignore some warning logs


from keras.models import Sequential
from keras.layers import Dense, Dropout


#  Define a Feed-Forward Model with 2 hidden layers with dimensions 392 and 196 Neurons
model_2 = Sequential([
  Dense(100, activation='relu', input_shape=(113,), name="first_hidden_layer"),
  Dense(500, activation='relu', name="second_hidden_layer"), Dropout(0.5),
  Dense(500,activation= 'relu',name = "third_hidden_layer"),
  Dense(500,activation= 'relu',name = "forth_hidden_layer"),
  Dense(100,activation= 'relu',name = "fifth_hidden_layer"),
  Dense(100,activation= 'relu',name = "sixth_hidden_layer"),
  Dense(5, activation='softmax'),
])

#  Validate your Model Architecture
print(model_2.summary())

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
first_hidden_layer (Dense)   (None, 100)               11400     
_________________________________________________________________
second_hidden_layer (Dense)  (None, 500)               50500     
_________________________________________________________________
dropout_1 (Dropout)          (None, 500)               0         
_________________________________________________________________
third_hidden_layer (Dense)   (None, 500)               250500    
_________________________________________________________________
forth_hidden_layer (Dense)   (None, 500)               250500    
_________________________________________________________________
fifth_hidden_layer (Dense)   (None, 100)               50100     
_________________________________________________________________
sixth_hidden_layer (Dense)   (None, 100)              

In [233]:
import keras_metrics as km
# Compile model
model_2.compile(optimizer='sgd', loss='categorical_crossentropy',metrics=['accuracy',km.binary_precision(), km.binary_recall()])

# Train model
model_2.fit(X_train_placed, y_train_placed_cat, epochs=100, batch_size=10,)

tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> tp
tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> fp
tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> tp
tracking <tf.Variable 'Variable:0' shape=() dtype=int32, numpy=0> fn
Epoch 1/100
860/860 [==============================] - 2s 2ms/step - loss: 1.6693 - accuracy: 0.3000 - precision: 0.3260 - recall: 0.1291
Epoch 2/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4990 - accuracy: 0.3267 - precision: 0.3260 - recall: 0.0840
Epoch 3/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4857 - accuracy: 0.3198 - precision: 0.1760 - recall: 0.0092
Epoch 4/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4783 - accuracy: 0.3314 - precision: 0.3541 - recall: 0.0317
Epoch 5/100
860/860 [==============================] - 1s 998us/step - loss: 1.4785 - accuracy: 0.3116 - precision: 0.5873 - recall: 0.0228
Epoch 6/100
860/860 [===========

860/860 [==============================] - 1s 1ms/step - loss: 1.4474 - accuracy: 0.3302 - precision: 0.6837 - recall: 0.0234
Epoch 57/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4481 - accuracy: 0.3384 - precision: 0.8143 - recall: 0.0315
Epoch 58/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4484 - accuracy: 0.3477 - precision: 0.5879 - recall: 0.0253
Epoch 59/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4469 - accuracy: 0.3453 - precision: 0.4694 - recall: 0.0160
Epoch 60/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4498 - accuracy: 0.3279 - precision: 0.3622 - recall: 0.0243
Epoch 61/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4497 - accuracy: 0.3267 - precision: 0.6548 - recall: 0.0502
Epoch 62/100
860/860 [==============================] - 1s 1ms/step - loss: 1.4488 - accuracy: 0.3337 - precision: 0.8508 - recall: 0.0302
Epoch 63/100
860/860 [==================

### The above model is not useful in predicting how long it will take for a fellow to get placed. One main reason is that for neural networks, we need more data to get a good model with neural networks, also if recall is bad, model is also bad and recall of this model is horrendous.

In [234]:
# Evaluate your model's performance on the test data
performance_test_model_2 = model_2.evaluate(X_test_placed, y_test_placed_cat)
print("Accuracy on Test samples: {0}".format(performance_test_model_2[1]))

96/96 [==============================] - 0s 4ms/step
Accuracy on Test samples: 0.25


In [244]:
from sklearn import tree
clf_placed = tree.DecisionTreeClassifier(max_depth = 100,random_state=300)
clf_placed = clf_placed.fit(X_train_placed, Y_train_placed)

y_pred_test_placed = clf_placed.predict(X_test_placed)
print("Accuracy with decision tree approach for test set = ",accuracy(Y_test_placed,y_pred_test_placed))

Accuracy with decision tree approach for test set =  0.28125


### Above model doesn't give a good accuracy

### SVM with hyper parameter tuning

In [245]:
from sklearn import svm

C_list = [0.8,1,1.5,10]
kernels = ['rbf', 'sigmoid']
for kernel in kernels:
  for c in C_list:
    #print(c)
    clf_svm = svm.SVC(C = c,kernel=kernel,random_state=20)
    clf_svm.fit(X_train_placed,Y_train_placed)
    y_pred_placed_svm = clf_svm.predict(X_test_placed)
    acc = accuracy(Y_test_placed,y_pred_placed_svm)
    print("c = ",c,"kernel = ",kernel,"acc = ",acc)

c =  0.8 kernel =  rbf acc =  0.3125
c =  1 kernel =  rbf acc =  0.3125
c =  1.5 kernel =  rbf acc =  0.28125
c =  10 kernel =  rbf acc =  0.3020833333333333
c =  0.8 kernel =  sigmoid acc =  0.2604166666666667
c =  1 kernel =  sigmoid acc =  0.2604166666666667
c =  1.5 kernel =  sigmoid acc =  0.3645833333333333
c =  10 kernel =  sigmoid acc =  0.3541666666666667


### Random Forest Classifier

In [246]:
from sklearn.ensemble import RandomForestClassifier

#max_depth = [2,3,4,5,6]

clf_rf = RandomForestClassifier(max_depth=150, random_state=500)
clf_rf.fit(X_train_placed, Y_train_placed)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=150, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=500,
                       verbose=0, warm_start=False)

In [247]:
y_pred_test_placed_rf = clf_rf.predict(X_test_placed)
print("Accuracy with decision tree approach for test set = ",accuracy(Y_test_placed,y_pred_test_placed_rf))

Accuracy with decision tree approach for test set =  0.3541666666666667


### Feature Selection

In [260]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.datasets import load_iris
from sklearn.feature_selection import SelectFromModel
clf_fs = ExtraTreesClassifier(n_estimators=250)
clf_fs = clf_fs.fit(X_placed, Y_placed)
#clf_fs.feature_importances

In [261]:
model = SelectFromModel(clf_fs, prefit=True)
X_new = model.transform(X_placed)

In [262]:
X_new.shape

(956, 37)

In [263]:
X_train_new_placed, X_test_new_placed, Y_train_new_placed, Y_test_new_placed = train_test_split(X_new, Y_placed, test_size=0.1, random_state=4)
print('X_train_placed_shape = ', X_train_new_placed.shape)
print('X_test_placed_shape = ', X_test_new_placed.shape)

X_train_placed_shape =  (860, 37)
X_test_placed_shape =  (96, 37)


In [264]:
from sklearn.ensemble import RandomForestClassifier

#max_depth = [2,3,4,5,6]

clf_rf = RandomForestClassifier(max_depth=50, random_state=100)
clf_rf.fit(X_train_new_placed, Y_train_new_placed)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=50, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=100,
                       verbose=0, warm_start=False)

In [265]:
y_pred_test_placed_rf = clf_rf.predict(X_test_new_placed)
print("Accuracy with decision tree approach for test set = ",accuracy(Y_test_new_placed,y_pred_test_placed_rf))

Accuracy with decision tree approach for test set =  0.3125


### Feature transformation and reduction - PCA

In [266]:
from sklearn.decomposition import PCA
pca = PCA(n_components=50)
X_placed_pca = pca.fit_transform(X_placed)

In [267]:
X_placed_pca.shape

(956, 50)

In [268]:
X_train_pca_placed, X_test_pca_placed, Y_train_pca_placed, Y_test_pca_placed = train_test_split(X_placed_pca, Y_placed, test_size=0.1, random_state=4)
print('X_train_pca_placed = ', X_train_pca_placed.shape)
print('X_test_pca_placed = ', X_test_pca_placed.shape)

X_train_pca_placed =  (860, 50)
X_test_pca_placed =  (96, 50)


In [269]:
from sklearn import svm

C_list = [0.8,1,1.5,10]
kernels = ['rbf', 'sigmoid']
for kernel in kernels:
  for c in C_list:
    #print(c)
    clf_svm = svm.SVC(C = c,kernel=kernel,random_state=20)
    clf_svm.fit(X_train_pca_placed,Y_train_pca_placed)
    y_pred_placed_pca_svm = clf_svm.predict(X_test_pca_placed)
    acc = accuracy(Y_test_pca_placed,y_pred_placed_pca_svm)
    #c_dict[acc] = [c,kernel]
    print("c = ",c,"kernel = ",kernel,"acc = ",acc)

c =  0.8 kernel =  rbf acc =  0.3020833333333333
c =  1 kernel =  rbf acc =  0.3020833333333333
c =  1.5 kernel =  rbf acc =  0.3229166666666667
c =  10 kernel =  rbf acc =  0.28125
c =  0.8 kernel =  sigmoid acc =  0.3020833333333333
c =  1 kernel =  sigmoid acc =  0.3020833333333333
c =  1.5 kernel =  sigmoid acc =  0.3020833333333333
c =  10 kernel =  sigmoid acc =  0.2916666666666667


### With the available data, 960 data points for placed candidates is pretty less. The best way to improve accuracy is by using feature selection, feature transformation, ensemble learning, trying various algorithms. After trying all the approaches, the best model is the one that gives 36% accuracy(SVM) on determing how long it will take for a person to get placed. The only way now to improve accuracy is to collect more data. Also, a regression approach is a better way to find how long it would take for a get a job. For this, instead of mentioning ranges of months, exact duration would be helpful in determining the duration. Better features could be included like whether the resume was reviewed, whether there were previous projects in the field of job applied, whether the candidate prepares before interview, how long do they revise, etc.,